ext/pdo/pdo_dbh.c

```diff
@@ -817,17 +810,18 @@ static bool pdo_dbh_attribute_set(pdo_dbh_t *dbh, zend_long attr, zval *value) /
 				return false;
 			}
 			dbh->def_stmt_ce = pce;
-			if (!Z_ISUNDEF(dbh->def_stmt_ctor_args)) {
-				zval_ptr_dtor(&dbh->def_stmt_ctor_args);
-				ZVAL_UNDEF(&dbh->def_stmt_ctor_args);
+			if (dbh->def_stmt_ctor_args) {
+				zend_array_release(dbh->def_stmt_ctor_args);
+				dbh->def_stmt_ctor_args = NULL;
 			}
 			if ((item = zend_hash_index_find(Z_ARRVAL_P(value), 1)) != NULL) {
 				if (Z_TYPE_P(item) != IS_ARRAY) {
 					zend_type_error("PDO::ATTR_STATEMENT_CLASS constructor_args must be of type ?array, %s given",
 						zend_zval_value_name(value));
 					return false;
 				}
-				ZVAL_COPY(&dbh->def_stmt_ctor_args, item);
+				/* Separate array */
+				dbh->def_stmt_ctor_args = zend_array_dup(Z_ARRVAL_P(item));
```

Comment:
Why do you need to duplicate the array? Previously we just incremented reference counter.

Reply:
Looks like I was double freeing the HashTable, once manually in the free handler, and another time indirectly via the GC. Let me know if the new code still makes sense.